src/parser/SGTypes.ts

```diff
@@ -11,19 +10,152 @@ export interface SGToken {
+                this.tokens.key,
+                this.tokens.equals,
+                this.tokens.openingQuote,
+                this.tokens.value,
+                this.tokens.closingQuote
+            );
+        }
+        return this._range;
+    }
+    private _range = null as Range;
+
+    public transpile(state: TranspileState) {
+        const result = [
+            state.transpileToken(this.tokens.key)
+        ];
+        if (this.tokens.value) {
+            result.push(
+                state.transpileToken(this.tokens.equals, '='),
+                state.transpileToken(this.tokens.openingQuote, '"'),
+                state.transpileToken(this.tokens.value),
+                state.transpileToken(this.tokens.closingQuote, '"')
+            );
+        }
+        return new SourceNode(null, null, null, result);
+    }
 }
 
 export class SGTag {
 
     constructor(
-        public tag: SGToken,
-        public attributes: SGAttribute[] = [],
-        public range?: Range
-    ) { }
+        startTagOpen: SGToken,
+        startTagName: SGToken,
+        attributes = [] as SGAttribute[],
+        startTagClose?: SGToken,
+        childNodes = [] as SGTag[],
+        endTagOpen?: SGToken,
+        endTagName?: SGToken,
+        endTagClose?: SGToken
+    ) {
+        this.tokens.startTagOpen = startTagOpen;
+        this.tokens.startTagName = startTagName;
+        this.attributes = attributes;
+        this.tokens.startTagClose = startTagClose;
+        this.childNodes = childNodes;
+        this.tokens.endTagOpen = endTagOpen;
+        this.tokens.endTagName = endTagName;
+        this.tokens.endTagClose = endTagClose;
+    }
+
+    public tokens = {} as {
+        /**
+        * The first portion of the startTag. (i.e. `<` or `<?`)
+        */
+        startTagOpen: SGToken;
+        /**
+         * The name of the opening tag (i.e. CoolTag in `<CoolTag>`).
+         */
+        startTagName: SGToken;
+        /**
+         * The last bit of the startTag (i.e. `/>` for self-closing, `?>` for xml prologue, or `>` for tag with children)
+         */
+        startTagClose?: SGToken;
+        /**
+         * The endTag opening char `<`
+         */
+        endTagOpen?: SGToken;
+        /**
+         * The name of the ending tag (i.e. CoolTag in `</CoolTag>`)
+         */
+        endTagName?: SGToken;
+        /**
+         * The endTag closing char `>`
+         */
+        endTagClose?: SGToken;
+    };
+
+    /**
+     * Array of attributes found on this tag
+     */
+    public attributes = [] as SGAttribute[];
+
+    /**
+     * The array of direct children AST elements of this AST node
+     */
+    public childNodes = [] as SGTag[];
+
+    public get range() {
+        if (!this._range) {
+            this._range = util.createBoundingRange(
+                this.tokens.startTagOpen,
+                this.tokens.startTagName,
+                this.attributes?.[this.attributes?.length - 1],
```

Comment:
Could we simplify by assuming `attributes` and `childNodes` should never be null?

Reply:
Yeah, good point. I worry about plugin authors creating bad AST Nodes. However, I'll update the constructor function to set these to empty array if passed in as null, so that should be safe enough.